common/src/main/java/dev/latvian/mods/kubejs/util/ModResourceBindings.java

```diff
@@ -0,0 +1,207 @@
+					.toList();
+			bindings.putAll(mod.getModId(), providers);
+		}
+	}
+
+	private String removeComment(String line) {
+		int index = line.indexOf('#');
+		return index == -1 ? line : line.substring(0, index);
+	}
+
+	@Nullable
+	private BindingProvider createProvider(Mod mod, String line) {
+		// SERVER name class field/method/<init>?
+		String[] split = line.split(" +");
+		if (split.length < 3) {
+			KubeJS.LOGGER.error("Invalid binding for '{}' in line: {}", mod.getModId(), line);
+			return null;
+		}
+
+		var scriptTypeFilter = createScriptTypeFilter(split[0]);
+		var name = split[1];
+		var className = split[2];
+		ClassBindingProvider classProvider = new ClassBindingProvider(name, scriptTypeFilter, className);
+		if (split.length == 3) {
+			return classProvider;
+		}
+
+		var methodFieldOrConstructor = split[3];
+		if (methodFieldOrConstructor.equals("<init>")) {
+			return new InstanceBindingProvider(classProvider);
+		}
+
+		return new InvokeBindingProvider(classProvider, methodFieldOrConstructor);
+	}
+
+	private Predicate<ScriptType> createScriptTypeFilter(String potentialType) {
+		if (potentialType.equals("*")) {
+			return type -> true;
+		}
+
+		ScriptType type = ScriptType.valueOf(potentialType.toUpperCase());
+		return type::equals;
+	}
+
+	interface BindingProvider extends Predicate<ScriptType> {
+
+		String name();
+
+		Object generate();
+	}
+
+	record ClassBindingProvider(String name, Predicate<ScriptType> filter, String className) implements BindingProvider {
+
+		@Override
+		public Object generate() {
+			try {
+				return this.getClass().getClassLoader().loadClass(className);
+			} catch (ClassNotFoundException e) {
+				throw new RuntimeException(e);
+			}
+		}
+
+		@Override
+		public boolean test(ScriptType scriptType) {
+			return filter.test(scriptType);
+		}
+	}
+
+	record InstanceBindingProvider(ClassBindingProvider parent) implements BindingProvider {
+
+		@Override
+		public String name() {
+			return parent.name();
+		}
+
+		@Override
+		public Object generate() {
+			Class<?> clazz = (Class<?>) parent().generate();
+			try {
+				Constructor<?> constructor = clazz.getConstructor();
+				return constructor.newInstance();
+			} catch (NoSuchMethodException | InvocationTargetException | InstantiationException | IllegalAccessException e) {
+				throw new IllegalStateException("[Bindings] Failed to find default constructor in class '" + clazz.getName() + "'");
+			}
+		}
+
+		@Override
+		public boolean test(ScriptType scriptType) {
+			return parent.test(scriptType);
+		}
+	}
+
+	record InvokeBindingProvider(ClassBindingProvider parent, String methodOrField) implements BindingProvider {
+
+		@Override
+		public String name() {
+			return parent().name();
+		}
+
+		@Override
+		public Object generate() {
+			Class<?> clazz = (Class<?>) parent().generate();
+
+			Object f = byField(clazz);
+			if (f != null) {
+				return f;
+			}
+
+			Object m = byMethod(clazz);
+			if (m != null) {
+				return m;
+			}
```

Comment:
Shouldn't this return a method ref instead of invoking the method, or am I mistaken here?

Reply:
The invoke serves as factory to load additional stuff before. 

```java
public static Object create() {
  var config = readConfig()
  return createConfigBinding(config);
}
```

Reply:
I... don't really see the usecase of that, sorry, maybe someone else can enlighten me

Reply:
What usecase exactly you don't see here?